zebrad/src/components/inbound.rs

```diff
@@ -111,6 +111,73 @@ pub struct InboundSetupData {
     pub latest_chain_tip: zs::LatestChainTip,
 }
 
+/// Caches and refreshes a partial list of peer addresses to be returned as a `GetAddr` response.
+pub struct CachedPeerAddrs {
+    /// A shared list of peer addresses.
+    address_book: Arc<Mutex<zn::AddressBook>>,
+
+    /// An owned list of peer addresses used as a `GetAddr` response.
+    cached_addrs: Vec<MetaAddr>,
+
+    /// Instant after which `cached_addrs` should be refreshed.
+    refresh_time: Instant,
+}
+
+impl CachedPeerAddrs {
+    /// Creates a new [`CachedPeerAddrs`].
+    fn new(address_book: Arc<Mutex<AddressBook>>) -> Self {
+        let cached_addrs = address_book
+            .lock()
+            .expect("previous thread panicked while holding the address book lock")
+            .sanitized_window();
+
+        let refresh_time = Instant::now() + INBOUND_CACHED_ADDRS_REFRESH_INTERVAL;
+
+        Self {
+            address_book,
+            cached_addrs,
+            refresh_time,
+        }
+    }
+
+    /// Refreshes the `cached_addrs` if the time has past `refresh_time`
+    fn try_refresh(&mut self) {
+        let now = Instant::now();
+
+        // return early if the cached addresses are still fresh
+        if now < self.refresh_time {
+            return;
+        }
+
+        // try getting a lock on the address book if it's time to refresh the cached addresses
+        match self.address_book.try_lock() {
+            Ok(address_book) => {
+                self.cached_addrs = address_book.sanitized_window();
+                self.refresh_time = now + INBOUND_CACHED_ADDRS_REFRESH_INTERVAL;
+            }
+
+            Err(TryLockError::WouldBlock) => {
+                let next_refresh_time = self.refresh_time + INBOUND_CACHED_ADDRS_REFRESH_INTERVAL;
+
+                if now > next_refresh_time {
+                    warn!("getaddrs response hasn't been refreshed in some time");
+                };
```

Comment:
I think the only way this could happen currently would be by queuing RPC requests to hold onto the lock, which isn't a concern. But yeah, can't hurt if something changes.

Reply:
It could also happen if Zebra is under load, and peers rapidly connect and disconnect. We have rate-limits, but for outbound connections they are still quite fast.

Reply:
> It could also happen if Zebra is under load, and peers rapidly connect and disconnect. We have rate-limits, but for outbound connections they are still quite fast.

You're right, I think that's worth "fixing", I'll open a separate PR to await the lock and a candidate outside the spawned tokio thread herre https://github.com/ZcashFoundation/zebra/blob/rate-limit-getaddr/zebra-network/src/peer_set/initialize.rs#L881

Reply:
That fix isn't needed, because the mutex is futures-aware.